reduce compiler `Assemble` complexity

onur-ozkan opened the pull request:
`compile::Assemble` is already complicated by its nature (as it handles core internals like recursive building logic, etc.) and also handles half of `LldWrapper` tool logic for no good reason since it should be done in the build step directly.

This change moves it there to reduce complexity of `compile::Assemble` logic.

<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r\? <reviewer name> (with the `\` removed)
-->
<!-- homu-ignore:end -->

jieyouxu commented on src/bootstrap/src/core/build_steps/tool.rs:
```diff
@@ -744,7 +748,26 @@ impl Step for LldWrapper {
-        })
+        });
+
+        let libdir = builder.sysroot_target_libdir(
+            Compiler { stage: builder.top_stage, host: self.target },
+            self.target,
+        );
+        let libdir_bin = libdir.parent().unwrap().join("bin");
```
Question: can this use `sysroot_target_bindir`?

onur-ozkan replied:
Yeap, I forgot that actually exist.

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12378612595/job/34550906298) [(plain)](https://github.com/rust-lang/rust/commit/ded8a6fdc37d4ba7c60ee8f065e680d7e8069e2f/checks/34550906298/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

onur-ozkan committed:
reduce compiler `Assemble` complexity

`compile::Assemble` is already complicated by its nature (as it handles core
internals like recursive building logic, etc.) and also handles half of `LldWrapper`
tool logic for no good reason since it should be done in the build step directly.

This change moves it there to reduce complexity of `compile::Assemble` logic.

Signed-off-by: onur-ozkan <work@onurozkan.dev>

jieyouxu commented on src/bootstrap/src/core/build_steps/tool.rs:
```diff
@@ -744,7 +748,25 @@ impl Step for LldWrapper {
+        });
+
+        let libdir_bin = builder.sysroot_target_bindir(
+            Compiler { stage: self.compiler.stage + 1, host: self.target },
```
Question: this is off-by-one stage, is that intentional? If so, could you leave a comment for why?

jieyouxu reviewed (commented):
Thanks, one question regarding the compiler stage, but otherwise LGTM.

I tested the changes locally on `x86_64-unknown-linux-gnu` with `rust.lld = true` up to stage 2 rustc and ran the resulting rustc with `-vV`, didn't seem to explode, so AFAICT this is fine.

jieyouxu reviewed (approved):
Thanks, you can r=me after PR CI is green.

onur-ozkan commented:
@bors r=jieyouxu

bors commented:
:pushpin: Commit bb1a90f939cf07a717a36af60b9f58f90ede26c0 has been approved by `jieyouxu`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jieyouxu bb1a90f939cf07a717a36af60b9f58f90ede26c0 -->
<!-- homu: {"type":"Approved","sha":"bb1a90f939cf07a717a36af60b9f58f90ede26c0","approver":"jieyouxu","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
